File: Ruby/examples/http_utils_example.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# HTTP Utils Example
# Demonstrates usage of the HTTP utilities module

require_relative '../http_utils/mod'

puts "=" * 60
puts "HTTP Utils Examples"
puts "=" * 60

# ============================================================================
# Example 1: URL Encoding/Decoding
# ============================================================================
puts "\n--- Example 1: URL Encoding/Decoding ---"

original = "hello world! special chars: &=?#"
encoded = HttpUtils.url_encode(original)
decoded = HttpUtils.url_decode(encoded)

puts "Original: #{original}"
puts "Encoded:  #{encoded}"
puts "Decoded:  #{decoded}"

# ============================================================================
# Example 2: Building Query Strings
# ============================================================================
puts "\n--- Example 2: Building Query Strings ---"

params = {
  'search' => 'ruby programming',
  'page' => '1',
  'limit' => '10'
}

query_string = HttpUtils.build_query_string(params)
puts "Query string: #{query_string}"

# ============================================================================
# Example 3: Building URLs with Parameters
# ============================================================================
puts "\n--- Example 3: Building URLs with Parameters ---"

base_url = 'https://api.example.com/search'
url = HttpUtils.build_url(base_url, params)
puts "Full URL: #{url}"

# Adding to existing query
existing_url = 'https://api.example.com/search?sort=date'
new_url = HttpUtils.build_url(existing_url, { 'page' => '2' })
puts "Updated URL: #{new_url}"

# ============================================================================
# Example 4: Parsing URLs
# ============================================================================
puts "\n--- Example 4: Parsing URLs ---"

url = 'https://user:pass@api.example.com:8080/v1/users?id=123#section'
parsed = HttpUtils.parse_url(url)

if parsed
  puts "Parsed URL components:"
  puts "  Scheme:   #{parsed[:scheme]}"
  puts "  Host:     #{parsed[:host]}"
  puts "  Port:     #{parsed[:port]}"
  puts "  Path:     #{parsed[:path]}"
  puts "  Query:    #{parsed[:query]}"
  puts "  Fragment: #{parsed[:fragment]}"
  puts "  Userinfo: #{parsed[:userinfo]}"
else
  puts "Failed to parse URL"
end

# ============================================================================
# Example 5: Parsing Query Strings
# ============================================================================
puts "\n--- Example 5: Parsing Query Strings ---"

query = 'name=John+Doe&age=30&city=New+York'
parsed_query = HttpUtils.parse_query_string(query)

puts "Parsed query parameters:"
parsed_query.each do |key, value|
  puts "  #{key}: #{value}"
end

# ============================================================================
# Example 6: URL Validation
# ============================================================================
puts "\n--- Example 6: URL Validation ---"

urls_to_test = [
  'https://example.com',
  'http://localhost:3000',
  'ftp://files.example.com',
  'not a url',
  'example.com'
]

urls_to_test.each do |test_url|
  valid = HttpUtils.valid_url?(test_url)
  puts "  #{test_url.ljust(30)} -> #{valid ? 'Valid' : 'Invalid'}"
end

# ============================================================================
# Example 7: Extracting URL Components
# ============================================================================
puts "\n--- Example 7: Extracting URL Components ---"

test_urls = [
  'https://api.github.com/users/octocat',
  'https://www.example.com:8080/path/to/resource',
  'invalid url here'
]

test_urls.each do |test_url|
  domain = HttpUtils.get_domain(test_url)
  path = HttpUtils.get_path(test_url)
  puts "  URL: #{test_url}"
  puts "    Domain: #{domain || 'N/A'}"
  puts "    Path:   #{path || 'N/A'}"
end

# ============================================================================
# Example 8: HTTP Request Options
# ============================================================================
puts "\n--- Example 8: HTTP Request Options ---"

options = HttpUtils::Options.new
options.timeout = 60
options.headers = {
  'Authorization' => 'Bearer token123',
  'Accept' => 'application/json'
}
options.verify_ssl = true

puts "Request options:"
puts "  Timeout:          #{options.timeout} seconds"
puts "  Headers:          #{options.headers.inspect}"
puts "  Follow redirects: #{options.follow_redirects}"
puts "  Max redirects:    #{options.max_redirects}"
puts "  Verify SSL:       #{options.verify_ssl}"

# ============================================================================
# Example 9: Creating a Response Object (for testing)
# ============================================================================
puts "\n--- Example 9: Response Object ---"

response = HttpUtils::Response.new(
  status_code: 200,
  status_message: 'OK',
  headers: {
    'Content-Type' => 'application/json',
    'X-Request-ID' => 'abc123'
  },
  body: '{"status":"success","data":{"id":1,"name":"Test"}}',
  url: 'https://api.example.com/users',
  response_time: 150
)

puts "Response details:"
puts "  Status code:    #{response.status_code}"
puts "  Status message: #{response.status_message}"
puts "  Success:        #{response.success?}"
puts "  Response time:  #{response.response_time}ms"
puts "  Content-Type:   #{response.header('content-type')}"

if response.json?
  puts "  Parsed JSON:    #{response.json.inspect}"
end

# ============================================================================
# Example 10: Practical URL Manipulation
# ============================================================================
puts "\n--- Example 10: Practical URL Manipulation ---"

# Building a search URL
base = 'https://search.example.com'
search_params = {
  'q' => 'ruby http client',
  'category' => 'programming',
  'sort' => 'relevance',
  'page' => '1'
}

search_url = HttpUtils.build_url(base, search_params)
puts "Search URL: #{search_url}"

# Adding pagination
next_page_url = HttpUtils.add_query_params(search_url, { 'page' => '2' })
puts "Next page:  #{next_page_url}"

# ============================================================================
# Example 11: Working with Form Data
# ============================================================================
puts "\n--- Example 11: Simulating Form Submission ---"

form_data = {
  'username' => 'john_doe',
  'email' => 'john@example.com',
  'message' => 'Hello, this is a test message!'
}

# This is how post_form would encode the data
encoded_form = HttpUtils.build_query_string(form_data)
puts "Form data encoded: #{encoded_form}"

# ============================================================================
# Summary
# ============================================================================
puts "\n" + "=" * 60
puts "Examples completed!"
puts "=" * 60
puts "\nNote: Actual HTTP requests require network access."
puts "Use HttpUtils.get(), HttpUtils.post(), etc. for real requests."

File: Perl/examples/http_utils_example.pl
#!/usr/bin/perl
# HTTP Utilities Example
# Demonstrates usage of AllToolkit::HttpUtils

use strict;
use warnings;
use utf8;

use FindBin;
use lib "$FindBin::Bin/../http_utils";

use AllToolkit::HttpUtils qw(:all);

print "=" x 60 . "\n";
print "Perl HTTP Utilities - Usage Examples\n";
print "=" x 60 . "\n\n";

#==============================================================================
# Example 1: URL Encoding/Decoding
#==============================================================================

print "Example 1: URL Encoding/Decoding\n";
print "-" x 40 . "\n";

my $text = "hello world! special chars: + = & ?";
my $encoded = url_encode($text);
my $decoded = url_decode($encoded);

print "Original: $text\n";
print "Encoded:  $encoded\n";
print "Decoded:  $decoded\n\n";

#==============================================================================
# Example 2: Build Query String
#==============================================================================

print "Example 2: Build Query String\n";
print "-" x 40 . "\n";

my $params = {
    q => "perl programming",
    page => 1,
    limit => 20
};

my $query = build_query_string($params);
print "Query string: $query\n\n";

#==============================================================================
# Example 3: Build Full URL
#==============================================================================

print "Example 3: Build Full URL\n";
print "-" x 40 . "\n";

my $base_url = "https://api.example.com/search";
my $full_url = build_url($base_url, $params);
print "Base URL: $base_url\n";
print "Full URL: $full_url\n\n";

#==============================================================================
# Example 4: Parse URL
#==============================================================================

print "Example 4: Parse URL\n";
print "-" x 40 . "\n";

my $url_to_parse = "https://user:pass@api.example.com:8080/v1/users?page=1&limit=10#section";
my $parsed = parse_url($url_to_parse);

print "URL: $url_to_parse\n";
print "  Scheme:   $parsed->{scheme}\n";
print "  Host:     $parsed->{host}\n";
print "  Port:     $parsed->{port}\n";
print "  Path:     $parsed->{path}\n";
print "  Username: $parsed->{username}\n";
print "  Password: $parsed->{password}\n";
print "  Fragment: $parsed->{fragment}\n";
print "  Query params:\n";
for my $key (sort keys %{$parsed->{query_params}}) {
    print "    $key = $parsed->{query_params}{$key}\n";
}
print "\n";

#==============================================================================
# Example 5: URL Validation
#==============================================================================

print "Example 5: URL Validation\n";
print "-" x 40 . "\n";

my @urls = (
    "https://example.com",
    "http://localhost:3000",
    "not-a-valid-url",
    "ftp://files.example.com",
);

for my $test_url (@urls) {
    my $valid = is_valid_url($test_url) ? "Valid" : "Invalid";
    print "$test_url -> $valid\n";
}
print "\n";

#==============================================================================
# Example 6: JSON Encoding/Decoding
#==============================================================================

print "Example 6: JSON Encoding/Decoding\n";
print "-" x 40 . "\n";

my $user_data = {
    name => "John Doe",
    email => "john@example.com",
    age => 30,
    roles => ["admin", "user"]
};

my $json_string = json_encode($user_data);
print "JSON encoded:\n$json_string\n\n";

my $decoded_data = json_decode($json_string);
print "Decoded data:\n";
print "  Name:  $decoded_data->{name}\n";
print "  Email: $decoded_data->{email}\n";
print "  Age:   $decoded_data->{age}\n";
print "  Roles: " . join(", ", @{$decoded_data->{roles}}) . "\n\n";

#==============================================================================
# Example 7: Query Parameter Manipulation
#==============================================================================

print "Example 7: Query Parameter Manipulation\n";
print "-" x 40 . "\n";

my $original_url = "https://example.com?existing=value";
my $new_url = add_query_params($original_url, { new_param => "new_value" });
print "Original: $original_url\n";
print "With new: $new_url\n";

my $clean_url = remove_query_params($new_url, ["existing"]);
print "Removed:  $clean_url\n\n";

#==============================================================================
# Example 8: Domain and Path Extraction
#==============================================================================

print "Example 8: Domain and Path Extraction\n";
print "-" x 40 . "\n";

my $api_url = "https://api.github.com/users/octocat/repos";
print "URL: $api_url\n";
print "Domain: " . get_domain($api_url) . "\n";
print "Path:   " . get_path($api_url) . "\n\n";

#==============================================================================
# Example 9: HTTP Options
#==============================================================================

print "Example 9: HTTP Options\n";
print "-" x 40 . "\n";

my $http_options = AllToolkit::HttpOptions->new(
    timeout => 60,
    headers => {
        'Authorization' => 'Bearer token123',
        'X-API-Key' => 'secret'
    },
    follow_redirects => 1,
    max_redirects => 5,
    verify_ssl => 1
);

print "HTTP Options:\n";
print "  Timeout: $http_options->{timeout}s\n";
print "  Follow redirects: $http_options->{follow_redirects}\n";
print "  Max redirects: $http_options->{max_redirects}\n";
print "  Verify SSL: $http_options->{verify_ssl}\n";
print "  Headers:\n";
for my $key (keys %{$http_options->{headers}}) {
    print "    $key: $http_options->{headers}{$key}\n";
}
print "\n";

#==============================================================================
# Example 10: HTTP Response Object
#==============================================================================

print "Example 10: HTTP Response Object\n";
print "-" x 40 . "\n";

my $mock_response = AllToolkit::HttpResponse->new(
    status_code    => 200,
    status_message => 'OK',
    body           => '{"id": 1, "name": "Test User"}',
    headers        => {
        'Content-Type' => 'application/json',
        'X-Request-ID' => 'abc123'
    },
    url            => 'https://api.example.com/users/1',
    success        => 1,
    response_time  => 0.234
);

print "Response:\n";
print "  Status:  $mock_response->{status_code} $mock_response->{status_message}\n";
print "  Success: $mock_response->{success}\n";
print "  Time:    " . sprintf("%.3f", $mock_response->{response_time}) . "s\n";
print "  URL:     $mock_response->{url}\n";
print "  Body:    $mock_response->{body}\n";

if ($mock_response->is_json()) {
    my $data = $mock_response->json();
    print "  Parsed:  ID=$data->{id}, Name=$data->{name}\n";
}
print "\n";

#==============================================================================
# Example 11: Complete API Request Simulation
#==============================================================================

print "Example 11: Complete API Request Simulation\n";
print "-" x 40 . "\n";

print "# This would make an actual HTTP request:\n";
print "# my \$response = http_get('https://api.example.com/users');\n";
print "#\n";
print "# POST JSON data:\n";
print "# my \$response = http_post_json('https://api.example.com/users', {\n";
print "#     name => 'John',\n";
print "#     email => 'john\@example.com'\n";
print "# });\n";
print "#\n";
print "# POST Form data:\n";
print "# my \$response = http_post_form('https://api.example.com/login', {\n";
print "#     username => 'admin',\n";
print "#     password => 'secret'\n";
print "# });\n";
print "\n";

print "=" x 60 . "\n";
print "Examples completed!\n";
print "=" x 60 . "\n";
